### PR TITLE
First check in for Migration Dashboard

### PR DESCRIPTION
**Pull request by aasimkhan30**
Database migration status is still work in Progress
image

This entire dialog will be refreshed. 
image

### CODE REVIEW

File: extensions/sql-migration/package.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
     "azdata": ">=1.19.0"
   },
   "activationEvents": [
+    "*",
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/package.json:15]: 1. If you have a * activation event then you don't need any others
2. Why is this necessary? We have a `onDashboardOpen` event if this was for the new dashboard page you added

File: extensions/sql-migration/package.json [hunk 2]
```diff
@@ -41,6 +42,35 @@
...
+      {
+        "id": "migration-dashboard",
+        "description": "%migration-dashboard-title%",
+        "provider": "MSSQL",
+        "title": "%migration-dashboard-title%",
+        "icon": {
+          "light": "./images/migration.svg",
+          "dark": "./images/migration.svg"
+        },
+        "when": "connectionProvider == 'MSSQL' && !mssql:iscloud",
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/package.json:56]: What about other ones like SQL on demand?

File: extensions/sql-migration/src/api/azure.ts [hunk 3]
```diff
@@ -127,8 +127,8 @@ export async function getMigrationControllerAuthKeys(account: azdata.Account, su
 		throw new Error(response.errors.toString());
 	}
 	return {
-		keyName1: response?.response?.data?.keyName1 ?? '',
-		keyName2: response?.response?.data?.keyName2 ?? ''
+		keyName1: response?.response?.data?.authKey1 ?? '',
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/api/azure.ts:130]: Do you want to rename the property names as well then?

File: extensions/sql-migration/src/api/azure.ts [hunk 4]
```diff
@@ -166,12 +166,24 @@ export async function startDatabaseMigration(account: azdata.Account, subscripti
 		throw new Error(response.errors.toString());
 	}
 	return {
-		errors: response.errors,
 		status: response.response.status,
 		databaseMigration: response.response.data
 	};
 }
 
+export async function getMigrationStatus(account: azdata.Account, subscription: Subscription, migration: DatabaseMigration): Promise<any> {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/api/azure.ts:174]: You should have an actual return type here - not `any`

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+const maxWidth = 800;
+
+export class DashboardWidget {
+
+	private _migrationStatusCardsContainer!: azdata.FlexContainer;
+	private _view!: azdata.ModelView;
+	/**
+	 * Creates new instance of dashboard
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:28]: This comment isn't really accurate - this doesn't actually create anything with just the constructor call

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+	 */
+	constructor(private _root: string) {
+	}
+
+	public register(): Promise<void> {
+		//Migrations.clearMigrations();
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:34]: Remove?

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+				'width': `${maxWidth}px`,
+				'padding-left': '20px'
+			}
+		});
+
+		return header;
+	}
+
+	private asAbsolutePath(filePath: string): string {
+		return path.join(this._root || '', filePath);
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:115]: If this can be undefined then the parameter should reflect that. If it can't then the || is unnecessary

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+			command: 'sqlmigration.start'
+		};
+		const migrateButton = this.createTaskButton(view, migrateButtonMetadata);
+
+		const preRequisiteListHTMLTemplate =
+			`
+			<span style="font-size: 14px;">Things you need before starting migration:</span><br>
+			<ul style="padding:0 0 0 10px;font-size:12px;">
+			<li>Azure account details</li>
+			<li>Azure SQL Managed Instance or SQL Server on Azure Virtual Machine</li>
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:141]: Localize these?

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+			<li>Backup location details</li>
+			</ul>
+			`;
+
+		const preRequisiteListElement = view.modelBuilder.dom().withProps({
+			html: preRequisiteListHTMLTemplate
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:147]: @alanrenmsft Do you know why we have this? This is dangerous for extensions to have direct access to...

@aasimkhan30 What can't you use direct modelview components for with this?

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+		}).component();
+
+		const preRequisiteLearnMoreLink = view.modelBuilder.hyperlink().withProps({
+			label: loc.LEARN_MORE,
+			url: '',
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:152]: I assume this will be added later? Make sure you're tracking stuff like this somewhere - it's easy to miss stuff like this

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+			height: maxHeight,
+			iconHeight: 32,
+			iconPath: taskMetaData.iconPath,
+			iconWidth: 32,
+			label: taskMetaData.title,
+			title: taskMetaData.title,
+			width: maxWidth,
+		}).component();
+		buttonContainer.onDidClick(async () => {
+			if (buttonContainer.enabled && taskMetaData.command) {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:201]: How could the button be clicked if it's disabled?

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+		return view.modelBuilder.divContainer().withItems([buttonContainer]).component();
+	}
+
+	private async refreshMigrations(): Promise<void> {
+		this._migrationStatusCardsContainer.clearItems();
+		const currentConnection = (await azdata.connection.getCurrentConnection());
+		const getMigrations = MigrationLocalStorage.getMigrations(currentConnection);
+		getMigrations.forEach((migration) => {
+			const button = this._view.modelBuilder.button().withProps({
+				label: `Migration to ${migration.targetManagedInstance.name} using controller ${migration.migrationContext.name}`
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:214]: localize

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+			CSSStyles: {
+				'margin-top': '30px'
+			}
+		});
+
+		return statusContainer;
+	}
+
+	private createVideoLinks(view: azdata.ModelView): azdata.Component {
+		//const maxWidth = 400;
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:340]: remove

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 14]
```diff
@@ -0,0 +1,510 @@
...
+			}
+		}).component();
+		video1Container.onDidClick(async () => {
+			if (linkMetaData.link) {
+				await vscode.env.openExternal(vscode.Uri.parse(linkMetaData.link));
+			}
+		});
+		videosContainer.addItem(video1Container, {
+			CSSStyles: {
+				'background-image': `url(${vscode.Uri.file(this.asAbsolutePath(linkMetaData.iconPath?.light as string || ''))})`,
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:499]: Typically we have an IconHelper class to provide icons so that we don't have to re-implement the same logic in every place that needs an icon

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+import { getMigrationStatus } from '../api/azure';
+import * as azdata from 'azdata';
+import { MigrationContext } from '../models/migrationLocalStorage';
+
+export class MigrationStatusDialog {
+
+	private _wizardObject: azdata.window.Dialog;
+
+	constructor(private migration: MigrationContext) {
+		this._wizardObject = azdata.window.createModelViewDialog('Migration Status', 'migrationStatusDialog', 'wide');
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dialog/migrationStatusDialog.ts:15]: localize title

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+			const statusText = view.modelBuilder.text().withProps({
+				value: JSON.stringify(status.result, undefined, 2)
+			}).component();
+
+			const loading = view.modelBuilder.loadingComponent().withProps({
+				loading: false
+			}).component();
+
+			const refreshButton = view.modelBuilder.button().withProps({
+				label: 'Refresh',
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dialog/migrationStatusDialog.ts:39]: localize

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+				width: '100px',
+				CSSStyles: {
+					'width': '30px'
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dialog/migrationStatusDialog.ts:42]: Why do you have two separate width styles here?

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+				}
+			}).component();
+
+			refreshButton.onDidClick(async (e) => {
+				loading.loading = true;
+				const status = await getMigrationStatus(this.migration.azureAccount, this.migration.subscription, this.migration.migrationContext);
+				statusHeader.value = status.result.properties.migrationStatus;
+				statusText.value = JSON.stringify(status.result, undefined, 2);
+				loading.loading = false;
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dialog/migrationStatusDialog.ts:51]: This should probably be in a finally block in case an error occurs

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+					{
+						component: statusText
+					},
+				],
+				{
+					horizontal: false
+				}
+			);
+			const form = formBuilder.withLayout({ width: '100%' }).component();
+			return view.initializeModel(form).then(() => {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dialog/migrationStatusDialog.ts:74]: Why the .then?

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 20]
```diff
@@ -465,8 +465,8 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 		);
 
 		console.log(response);
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/models/stateMachine.ts:467]: You should remove the (what I assume are) debug `console.log`s from checked in code so we don't have a bunch of that stuff cluttering up the console logs.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/sql-migration/src/api/azure.ts (hunk 4 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/api/azure.ts:174]: I will get around that in the next iteration. When I actually implement the migration status dialog

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts (hunk 18 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/dialog/migrationStatusDialog.ts:51]: This code is going to get removed in the next iteration. For now, I am just making the api call to check if the dialog opens correctly and I have the required data in there.

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts (hunk 14 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:499]: Yeah, I have already added that class. I will move it there.

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts (hunk 12 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:214]: This needs to be replaced. It is just a placeholder for now. I will localize the contents in the next iteration.

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts (hunk 19 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/dialog/migrationStatusDialog.ts:74]: Removing migration status dialog from this PR.

File: extensions/sql-migration/src/dialog/migrationStatusDialog.ts (hunk 18 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/dialog/migrationStatusDialog.ts:51]: Removing migration status dialog from this PR.

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts (hunk 9 above)
Comment (alanrenmsft) [reply extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:135]: this component is introduced by this PR: https://github.com/microsoft/azuredatastudio/pull/2622, Abbie didn't give description about why is this component needed, Matt did raise the same concern as you did in that PR. 

this issue is mentioned in this PR, so I guess it is for markdown related stuff? https://github.com/microsoft/azuredatastudio/issues/2626

@chlafreniere do you remember?

Comment (chlafreniere) [reply extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:135]: Pinged some details offline; this component was used in markdown rendering for a previous notebook implementation that is no longer needed any longer; we currently do markdown rendering in-proc.

Comment (aasimkhan30) [reply extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:135]: I have switched to text component now. 

Now:
image

Before:
image

File: extensions/sql-migration/src/api/azure.ts (hunk 4 above)
Comment (Charles-Gagnon) [reply extensions/sql-migration/src/api/azure.ts:174]: Is there any reason why you wouldn't want to do it now? Even if it's not the final form having something for the type will help ensure that everything is hooked up correctly.

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 21]
```diff
@@ -0,0 +1,502 @@
...
+			}
+		});
+
+		return tasksContainer;
+	}
+
+	private createTaskButton(view: azdata.ModelView, taskMetaData: IActionMetadata): azdata.Component {
+		const maxHeight: number = 84;
+		const maxWidth: number = 236;
+		const buttonContainer = view.modelBuilder.button().withProperties<azdata.ButtonProperties>({
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:185]: If you use withProps you don't need to specify the properties interface

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 22]
```diff
@@ -0,0 +1,502 @@
...
+
+export class DashboardWidget {
+
+	private _migrationStatusCardsContainer!: azdata.FlexContainer;
+	private _view!: azdata.ModelView;
+
+	constructor() {
+	}
+
+	public register(): Promise<void> {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:30]: From what I can tell nothing is actually waiting on the promise returned so I'd suggest just removing the promise and just have this function call registerModelViewProvider but not wait for it to be fully initialized. Having the stuff wrapped in the promise is just unnecessary currently.

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/sql-migration/images/background.svg [hunk 23]
```diff
@@ -0,0 +1,38 @@
+<svg width="753" height="242" viewBox="0 0 753 242" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (alanrenmsft) [inline extensions/sql-migration/images/background.svg:1]: could you please attach screenshots for dark and high contrast modes?

File: extensions/sql-migration/src/api/azure.ts [hunk 24]
```diff
@@ -166,12 +166,24 @@ export async function startDatabaseMigration(account: azdata.Account, subscripti
...
+	const api = await getAzureCoreAPI();
+	const host = `https://eastus2euap.management.azure.com`;
+	const path = `${migration.id}?$expand=MigrationStatusDetails&api-version=2020-09-01-preview`;
+	const response = await api.makeAzureRestRequest(account, subscription, path, azurecore.HttpRequestMethod.GET, undefined, true, host);
+	if (response.errors.length > 0) {
+		throw new Error(response.errors.toString());
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/api/azure.ts:180]: what does errors.toString() look like?

File: extensions/sql-migration/src/api/azure.ts [hunk 25]
```diff
@@ -166,12 +166,24 @@ export async function startDatabaseMigration(account: azdata.Account, subscripti
...
+	}
+	return {
+		errors: response.errors,
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/api/azure.ts:183]: not sure why do we need the errors property, it will never be actual errors.

File: extensions/sql-migration/src/api/azure.ts [hunk 26]
```diff
@@ -223,8 +235,8 @@ export interface MigrationController {
 }
 
 export interface GetMigrationControllerAuthKeysResult {
-	keyName1: string,
-	keyName2: string
+	authKey1: string,
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/api/azure.ts:238]: could you please add some comments why are there 2 auth keys in the result?

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 27]
```diff
@@ -0,0 +1,502 @@
...
+					height: '100%'
+				}).component();
+				const header = this.createHeader(view);
+
+				const tasksContainer = await this.createTasks(view);
+
+				container.addItem(header, {
+					CSSStyles: {
+						'background-image': `url(${IconPathHelper.migrationDashboardHeaderBackground})`,
+						'width': `1100px`,
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:46]: lets be consistent, I don't see a need to use ` here, you can use single quote.

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 28]
```diff
@@ -0,0 +1,502 @@
...
+
+		const migrateButtonMetadata: IActionMetadata = {
+			title: loc.DASHBOARD_MIGRATE_TASK_BUTTON_TITLE,
+			description: loc.DASHBOARD_MIGRATE_TASK_BUTTON_DESCRIPTION,
+			iconPath: IconPathHelper.sqlMigrationLogo,
+			command: 'sqlmigration.start'
+		};
+
+		const preRequisiteListTitle = view.modelBuilder.text().withProps({
+			value: `${loc.PRE_REQ_TITLE}`,
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:124]: no need to use a template string here

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 29]
```diff
@@ -0,0 +1,502 @@
...
+				'padding-top': '10px',
+				'height': '10px',
+				'margin': '0px'
+			}
+		});
+
+		const links = [{
+			title: loc.HELP_LINK1_TITLE,
+			description: loc.HELP_LINK1_DESCRIPTION,
+			link: 'www.microsoft.com'
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:367]: another todo here?

**Review by kburtram — APPROVED** [review]
Looks goods as a work in progress checkin.

File: extensions/sql-migration/src/api/azure.ts (hunk 24 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/api/azure.ts:180]: It just converts the Json errors to a string.

File: extensions/sql-migration/src/api/azure.ts (hunk 26 above)
Comment (aasimkhan30) [reply extensions/sql-migration/src/api/azure.ts:238]: Both the auth keys are for the same thing. There is no functional difference between the 2 authkeys. You can use either one of them.

File: extensions/sql-migration/images/background.svg (hunk 23 above)
Comment (aasimkhan30) [reply extensions/sql-migration/images/background.svg:1]: image

File: extensions/sql-migration/src/api/azure.ts (hunk 26 above)
Comment (Charles-Gagnon) [reply extensions/sql-migration/src/api/azure.ts:238]: Is there documentation you could link to that discusses this? I'm assuming there's two for key rotation purposes.

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 30]
```diff
@@ -0,0 +1,502 @@
...
+
+interface IActionMetadata {
+	title?: string,
+	description?: string,
+	link?: string,
+	iconPath?: { light: string | vscode.Uri; dark: string | vscode.Uri },
+	command?: string;
+}
+
+const maxWidth = 800;
```
[the remark this answers is not in the document]
Comment (aasimkhan30) [reply extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:20]: I think it comes from machine learning dashboard page.

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts [hunk 31]
```diff
@@ -27,55 +27,52 @@ export class DashboardWidget {
 	constructor() {
 	}
 
-	public register(): Promise<void> {
-		return new Promise<void>(resolve => {
-			azdata.ui.registerModelViewProvider('migration.dashboard', async (view) => {
-				this._view = view;
-				const container = view.modelBuilder.flexContainer().withLayout({
+	public register() {
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:30]: add return type `: void`

**Review by alanrenmsft — APPROVED** [review]

File: extensions/sql-migration/src/dashboard/sqlServerDashboard.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (aasimkhan30) [reply extensions/sql-migration/src/dashboard/sqlServerDashboard.ts:20]: I have tried it in a couple of resolutions and it seems fine. I will reiterate if there are some responsiveness bugs.
